File: app/src/main/java/net/bloople/audiobooks/IndexingTask.java
package net.bloople.audiobooks;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.media.MediaMetadataRetriever;
import android.os.AsyncTask;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class IndexingTask extends AsyncTask<String, Integer, Void> {
    public static final Pattern AUDIOBOOK_EXTENSIONS = Pattern.compile("\\.(mp3|m4a)$");

    private Context context;
    private Indexable indexable;
    private int progress;
    private int max;
    private int indexed;

    IndexingTask(Context context, Indexable indexable) {
        this.context = context;
        this.indexable = indexable;
    }

    protected Void doInBackground(String... params) {
        destroyDeleted();
        indexDirectory(new File(params[0]));
        publishProgress(progress, max);
        return null;
    }

    protected void onProgressUpdate(Integer... args) {
        indexable.onIndexingProgress(args[0], args[1]);
    }

    protected void onPostExecute(Void result) {
        indexable.onIndexingComplete(indexed);
    }

    private void destroyDeleted() {
        SQLiteDatabase db = DatabaseHelper.instance(context);

        Cursor cursor = db.query("books", null, null, null, null, null, null);

        max += cursor.getCount();

        while(cursor.moveToNext()) {
            Book book = new Book(cursor);

            File file = new File(book.path());
            if(!file.exists()) book.destroy(context);

            progress++;
            publishProgress(progress, max);
        }

        cursor.close();
    }

    private void indexDirectory(File directory) {
        File[] files = directory.listFiles();

        if(files == null) return;

        ArrayList<File> filesToIndex = new ArrayList<>();

        for(File f : files) {
            if(f.isDirectory()) {
                indexDirectory(f);
            }
            else {
                Matcher matcher = AUDIOBOOK_EXTENSIONS.matcher(f.getName());
                if(matcher.find()) filesToIndex.add(f);
            }
        }

        max += filesToIndex.size();
        publishProgress(progress, max);

        for(File f : filesToIndex) indexFile(f);
    }

    private void indexFile(File file) {
        try {
            Book book = Book.findByPath(context, file.getCanonicalPath());
            if(book == null) book = new Book();

            book.path(file.getCanonicalPath());
            Matcher matcher = AUDIOBOOK_EXTENSIONS.matcher(file.getName());
            book.title(matcher.replaceFirst(""));
            book.mtime(file.lastModified());
            book.size(getDuration(book.path()));


            book.save(context);
            progress++;
            indexed++;
        }
        catch(IOException e) {
            e.printStackTrace();
        }

        publishProgress(progress, max);
    }

    private int getDuration(String path) {
        try {
            MediaMetadataRetriever retriever = new MediaMetadataRetriever();
            retriever.setDataSource(path);
            String duration = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
            return Integer.parseInt(duration);
        }
        catch(RuntimeException e) {
            e.printStackTrace();
            return 0;
        }
    }
}

File: app/src/main/java/net/bloople/audiobooks/IndexingTask.kt
@file:Suppress("DEPRECATION")

package net.bloople.audiobooks

import android.content.Context
import android.os.AsyncTask
import java.io.File
import java.util.*

@Suppress("OVERRIDE_DEPRECATION")
internal class IndexingTask(private val context: Context, private val indexable: Indexable) :
    AsyncTask<String, Int, Unit>() {
    private var progress = 0
    private var max = 0
    private val metrics = IndexingMetrics()
    private val fileParser = FileParser()

    override fun doInBackground(vararg params: String) {
        destroyDeleted()
        fileParser.use { indexDirectory(File(params[0])) }
        publishProgress(progress, max)
    }

    override fun onProgressUpdate(vararg args: Int?) {
        indexable.onIndexingProgress(args[0]!!, args[1]!!)
    }

    override fun onPostExecute(result: Unit?) {
        indexable.onIndexingComplete(metrics)
    }

    private fun destroyDeleted() {
        Book.findAll(context) {
            max += it.count
            while (it.moveToNext()) {
                val book = Book(it)
                if (!book.file.exists()) {
                    book.destroy(context)
                    metrics.deleted++
                }

                progress++
                publishProgress(progress, max)
            }
        }
    }

    private fun indexDirectory(directory: File) {
        val (directories, files) = directory.listFiles()?.partition { it.isDirectory } ?: return
        for(d in directories) { indexDirectory(d) }

        files.filter { AUDIOBOOK_EXTENSIONS.contains(it.extension.lowercase()) }.also {
            max += it.size
            publishProgress(progress, max)
        }.forEach(::indexFile)
    }

    private fun indexFile(file: File) {
        val parsedFile by lazy { fileParser.parse(file) }
        val book = Book.findByPathOrNull(context, file.canonicalPath)
        if(book != null) updateBook(parsedFile, book)
        else createBook(parsedFile)

        progress++
        publishProgress(progress, max)
    }

    private fun createBook(parsedFile: ParsedFile) {
        NewBook(parsedFile.path, parsedFile.title, parsedFile.mtime, parsedFile.size).save(context)
        metrics.created++
    }

    private fun updateBook(parsedFile: ParsedFile, book: Book) {
        if(parsedFile.mtime == book.mtime) {
            metrics.skipped++
            return
        }

        book.edit(context) {
            title = parsedFile.title
            mtime = parsedFile.mtime
            size = parsedFile.size
        }
        metrics.updated++
    }

    companion object {
        val AUDIOBOOK_EXTENSIONS = listOf("m4a", "mkv", "mp3", "mp4", "wav", "webm")
    }
}